Source/WebCore/platform/text/cf/HyphenationCF.cpp

```diff
@@ -35,49 +35,29 @@
 namespace WTF {
 
 template<>
-class TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>
+RetainPtr<CFLocaleRef> TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>::createValueForKey(const AtomString& localeIdentifier)
 {
-public:
-    static TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>& cache()
-    {
-        static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
-        return cache;
-    }
-
-    static bool isKeyNull(const AtomString& localeIdentifier)
-    {
-        return localeIdentifier.isNull();
-    }
-
-    static RetainPtr<CFLocaleRef> createValueForNullKey()
-    {
-        // CF hyphenation functions use locale (regional formats) language, which doesn't necessarily match primary UI language,
-        // so we can't use default locale here. See <rdar://problem/14897664>.
-        RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, defaultLanguage().createCFString().get()));
-        return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
-    }
-
-    static RetainPtr<CFLocaleRef> createValueForKey(const AtomString& localeIdentifier)
-    {
-        RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, localeIdentifier.string().createCFString().get()));
-
-        return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
-    }
-
-    static AtomString createKeyForStorage(const AtomString& key) { return key; }
-};
+    RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, localeIdentifier.string().createCFString().get()));
+    return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
 }
 
+} // namespace WTF
+
 namespace WebCore {
 
+static RetainPtr<CFLocaleRef> cachedLocale(const AtomString& localeIdentifier)
+{
+    static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
+    return cache.get().get(localeIdentifier);
+}
 bool canHyphenate(const AtomString& localeIdentifier)
 {
-    return TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>::cache().get(localeIdentifier);
+    return cachedLocale(localeIdentifier);
```

Comment:
This change seems unnecessary.

Reply:
Since the templated `TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>` class was removed, we can't use the `static TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>& cache()` method anymore. A new static `cachedLocale()` function was added to create and access the` NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>>` cache. It's also shorter.

Reply:
Does the refactoring change behavior per se? Otherwise, maybe we should split this into 2 patches. One for refactoring and one for addressing the hyphenation bug, as it would be easier to follow.

Reply:
Why are we removing `TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>`? I don't think we should do that. It makes PR impossible to follow.

Reply:
Note that `TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>` is a **class** template specialization. If we keep using it we have to implement all 4 static methods, where 3 of them are almost identical to the templated implementation. Also, `isKeyNull()` and `createValueForNullKey()` are redundant since we don't store a locale for null `localeIdentifier`. It would look something like this.

```cpp
namespace WTF {

template<>
class TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>
{
public:
    static TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>& cache()
    {
        static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
        return cache;
    }

    static bool isKeyNull(const AtomString& localeIdentifier)
    {
        return localeIdentifier.isNull();
    }

    static RetainPtr<CFLocaleRef> createValueForNullKey()
    {
        return nullptr;
    }

    static RetainPtr<CFLocaleRef> createValueForKey(const AtomString& localeIdentifier)
    {
        RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, localeIdentifier.string().createCFString().get()));

        return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
    }

    static AtomString createKeyForStorage(const AtomString& key) { return key; }
};

} // namespace WTF
```
What I propose instead is to use a **function** template specialization and implement only `createValueForKey()` method:
```cpp
namespace WTF {

template<>
RetainPtr<CFLocaleRef> TinyLRUCachePolicy<AtomString, RetainPtr<CFLocaleRef>>::createValueForKey(const AtomString& localeIdentifier)
{
    RetainPtr<CFLocaleRef> locale = adoptCF(CFLocaleCreate(kCFAllocatorDefault, localeIdentifier.string().createCFString().get()));
    return CFStringIsHyphenationAvailableForLocale(locale.get()) ? locale : nullptr;
}

} // namespace WTF

namespace WebCore {

static RetainPtr<CFLocaleRef> cachedLocale(const AtomString& localeIdentifier)
{
    static NeverDestroyed<TinyLRUCache<AtomString, RetainPtr<CFLocaleRef>>> cache;
    return cache.get().get(localeIdentifier);
}

} // namespace WebCore
```

It's more compact and less confusing in my opinion.

This is similar to the `CGColor` cache in https://github.com/WebKit/WebKit/blob/e89434de872e7aa88c2d2119df0e30b1b8004d9e/Source/WebCore/platform/graphics/cg/ColorCG.cpp#L45-L53